### ISSUE TITLE
pilot groups not working

### ISSUE THREAD

**Issue by axi92** [fastlane/fastlane#21249]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
I use this to deploy my ipa to Testflight:
```
  upload_to_testflight(
    api_key_path: "xxxxxxxxx.json",
    app_identifier: "com.foo.bar",
    distribute_external: true,
    groups: ["SFW"] # ID for SFW
  )
```
I have to manually add them every time:
![image](https://user-images.githubusercontent.com/1394837/236418373-3aaf874b-9f0a-42cf-9757-f67099cbf457.png)

##### Command executed
bundle exec fastlane run <name of lane where this code is>

##### Complete output when running fastlane, including the stack trace and command used
The output

<details>
  <pre> [10:00:36]: [32m--- Step: upload_to_testflight ---[0m
[10:00:36]: [32m----------------------------------[0m
[10:00:36]: Using existing authorization token for App Store Connect API
[10:00:37]: [32mReady to upload new build to TestFlight (App: 6447353802)...[0m
[10:00:37]: Going to upload updated app to App Store Connect
[10:00:37]: [32mThis might take a few minutes. Please don't interrupt the script.[0m
[10:01:30]: [32m--------------------------------------------------------------------[0m
[10:01:30]: [32mSuccessfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.
[10:01:30]: [32m--------------------------------------------------------------------[0m
[10:01:30]: [32mSuccessfully uploaded the new binary to App Store Connect[0m
[10:01:30]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
[10:01:30]: Note that if `skip_waiting_for_build_processing` is used but a `changelog` is supplied, this process will wait ppStoreConnect, update the changelog and then skip the remaining of the processing steps.
[10:01:30]: Waiting for processing on... app_id: 6447353802, app_version: 0.0.1, build_version: 162, platform: IOS
[10:01:31]: [33mRead more information on why this build isn't showing up yet - https://github.com/fastlane/fastlane/issues/
[10:01:31]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for ntinues)
[10:02:01]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for ntinues)
[10:02:32]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for ntinues)
[10:03:02]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for ntinues)
[10:03:33]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for ntinues)
[10:04:03]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for ntinues)
[10:04:36]: Waiting for App Store Connect to finish processing the new build (0.0.1 - 162) for IOS
[10:05:07]: Waiting for App Store Connect to finish processing the new build (0.0.1 - 162) for IOS
[10:05:38]: Waiting for App Store Connect to finish processing the new build (0.0.1 - 162) for IOS
[10:06:08]: Waiting for App Store Connect to finish processing the new build (0.0.1 - 162) for IOS
[10:06:39]: Waiting for App Store Connect to finish processing the new build (0.0.1 - 162) for IOS
[10:07:10]: Waiting for App Store Connect to finish processing the new build (0.0.1 - 162) for IOS
[10:07:43]: Waiting for App Store Connect to finish processing the new build (0.0.1 - 162) for IOS
[10:08:16]: [32mSuccessfully finished processing the build 0.0.1 - 162 for IOS[0m
[10:08:16]: [33mUsing App Store Connect's default for notifying external testers (which is true) - set r full control[0m
[10:08:16]: Distributing new build to testers: 0.0.1 - 162
[10:08:17]: [33mExport compliance has been set to 'false'. Need to wait for build to finishing processing again...[0m
[10:08:17]: [33mSet 'ITSAppUsesNonExemptEncryption' in the 'Info.plist' to skip this step and speed up the submission[0m
[10:08:17]: [32mSuccessfully distributed build to Internal testers 🚀[0m

+------+------------------------+-------------+
|              [32mfastlane summary[0m               |
+------+------------------------+-------------+
| Step | Action                 | Time (in s) |
+------+------------------------+-------------+
| 1    | Switch to vault lane   | 0           |
| 2    | shell command          | 0           |
| 3    | Switch to              | 0           |
|      | get_provisioning_prof  |             |
|      | iles lane              |             |
| 4    | sigh                   | 1           |
| 5    | Switch to              | 0           |
|      | domain_sign_ipa lane     |             |
| 6    | mv ../*.ipa            | 0           |
|      | ../app.ipa             |             |
| 7    | resign                 | 2           |
| 8    | mv ../app.ipa          | 0           |
|      | ../app.iOS-signed  |             |
|      | -.ipa                  |             |
| 9    | upload_to_testflight   | 460         |
+------+------------------------+-------------+

[10:08:17]: [32mfastlane.tools just saved you 8 minutes! 🎉[0m
successfull published to testlfight</pre>
</details>

### Environment
[✔] 🚀 
[09:16:16]: Generating fastlane environment output, this might take a few seconds...
swift-driver version: 1.62.15 <details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                           |
| --------------------------- | ----------------------------------------------- |
| OS                          | 13.2.1                                          |
| Ruby                        | 3.2.2                                           |
| Bundler?                    | true                                            |
| Git                         | git version 2.37.1 (Apple Git-137.1)            |
| Installation Source         | ~/appname/vendor/bundle/ruby/3.2.0/bin/fastlane |
| Host                        | macOS 13.2.1 (22D68)                            |
| Ruby Lib Dir                | /opt/homebrew/Cellar/ruby/3.2.2/lib             |
| OpenSSL Version             | OpenSSL 1.1.1t  7 Feb 2023                      |
| Is contained                | false                                           |
| Is homebrew                 | false                                           |
| Is installed via Fabric.app | false                                           |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/     |
| Xcode Version               | 14.2                                            |
| Swift Version               | 5.7.2                                           |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
before_all do
  vault()
end

desc "Load secrets from vault"
lane :vault do
  sh("../bamboo-specs/vault.sh", log: false)
end


desc "Load provisioning profiles from apple server."
lane :get_provisioning_profiles do
  sigh(
    adhoc: true,
    app_identifier: "com.domain.appname",
    api_key_path: "xxxxxxx.json"
  )
  # add actions here: https://docs.fastlane.tools/actions
end

desc "Resign ipa file."
lane :domain_sign_ipa do
  sh("mv ../*.ipa ../app.ipa")
  # sh("security", "unlock-keychain", "-p", ENV["KEYCHAIN_PASS"], log: false)
  resign(
    ipa: "app.ipa",
    signing_identity: "Apple Distribution: domain (xxxxxxxxxx)",
    provisioning_profile: "AdHoc_com.domain.appname.mobileprovision"
  )
  sh("mv ../app.ipa ../appname.iOS-signed-.ipa")
end

desc "Upload ipa to testflight."
lane :deploy_to_testflight do
  upload_to_testflight(
    api_key_path: "xxxxxxx.json",
    app_identifier: "com.domain.appname",
    distribute_external: true,
    groups: ["SFW"] # ID for SFW
  )
end

desc "Run all lanes for testflight deployment"
lane :run_all_testflight do
  get_provisioning_profiles()
  domain_sign_ipa()
  upload_to_testflight()
end

desc "Run all lanes for google play deployment internal"
lane :run_all_google_internal do
  sh("mv ../*.aab ../app.aab")
  VERSION=sh('jq -r ".version" ../package.json')
  upload_to_play_store(
    package_name: "com.domain.appname",
    version_name: VERSION, # string
    version_code: ENV["bamboo_deploy_version"], #integer, the same can't be uploaded twice
    track: "internal",
    release_status: 'draft',
    json_key: "google_service.json"
  )
end
```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier("com.domain.appname") # The bundle identifier of your app
apple_id("u.ser@domain.com") # Your Apple email address
```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.212.2 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                             | Version      |
| ------------------------------- | ------------ |
| error_highlight                 | 0.5.1        |
| did_you_mean                    | 1.6.3        |
| syntax_suggest                  | 1.0.2        |
| bundler                         | 2.4.10       |
| pathname                        | 0.2.1        |
| rake                            | 13.0.6       |
| rexml                           | 3.2.5        |
| CFPropertyList                  | 3.0.6        |
| public_suffix                   | 5.0.1        |
| addressable                     | 2.8.4        |
| artifactory                     | 3.0.15       |
| atomos                          | 0.1.3        |
| aws-eventstream                 | 1.2.0        |
| aws-partitions                  | 1.749.0      |
| aws-sigv4                       | 1.5.2        |
| jmespath                        | 1.6.2        |
| aws-sdk-core                    | 3.171.0      |
| aws-sdk-kms                     | 1.63.0       |
| aws-sdk-s3                      | 1.120.1      |
| babosa                          | 1.0.4        |
| claide                          | 1.1.0        |
| colored                         | 1.2          |
| colored2                        | 3.1.2        |
| highline                        | 2.0.3        |
| commander                       | 4.6.0        |
| declarative                     | 0.0.20       |
| digest-crc                      | 0.6.4        |
| unf_ext                         | 0.0.8.2      |
| unf                             | 0.1.4        |
| domain_name                     | 0.5.20190701 |
| dotenv                          | 2.8.1        |
| emoji_regex                     | 3.2.3        |
| excon                           | 0.99.0       |
| faraday-em_http                 | 1.0.0        |
| faraday-em_synchrony            | 1.0.0        |
| faraday-excon                   | 1.1.0        |
| faraday-httpclient              | 1.0.1        |
| multipart-post                  | 2.0.0        |
| faraday-multipart               | 1.0.4        |
| faraday-net_http                | 1.0.1        |
| faraday-net_http_persistent     | 1.2.0        |
| faraday-patron                  | 1.0.0        |
| faraday-rack                    | 1.0.0        |
| faraday-retry                   | 1.0.3        |
| ruby2_keywords                  | 0.0.5        |
| faraday                         | 1.10.3       |
| http-cookie                     | 1.0.5        |
| faraday-cookie_jar              | 0.0.7        |
| faraday_middleware              | 1.2.0        |
| fastimage                       | 2.2.6        |
| gh_inspector                    | 1.1.3        |
| jwt                             | 2.7.0        |
| memoist                         | 0.16.2       |
| multi_json                      | 1.15.0       |
| os                              | 1.1.4        |
| signet                          | 0.17.0       |
| googleauth                      | 1.5.2        |
| httpclient                      | 2.8.3        |
| mini_mime                       | 1.1.2        |
| trailblazer-option              | 0.1.2        |
| uber                            | 0.1.0        |
| representable                   | 3.2.0        |
| retriable                       | 3.1.2        |
| webrick                         | 1.8.1        |
| google-apis-core                | 0.11.0       |
| google-apis-androidpublisher_v3 | 0.39.0       |
| google-apis-playcustomapp_v1    | 0.13.0       |
| google-apis-iamcredentials_v1   | 0.17.0       |
| google-apis-storage_v1          | 0.19.0       |
| google-cloud-env                | 1.6.0        |
| google-cloud-errors             | 1.3.1        |
| google-cloud-core               | 1.6.0        |
| google-cloud-storage            | 1.44.0       |
| json                            | 2.6.3        |
| mini_magick                     | 4.12.0       |
| naturally                       | 2.2.1        |
| optparse                        | 0.1.1        |
| plist                           | 3.7.0        |
| rubyzip                         | 2.3.2        |
| security                        | 0.1.3        |
| simctl                          | 1.6.10       |
| terminal-notifier               | 2.0.0        |
| unicode-display_width           | 1.8.0        |
| terminal-table                  | 1.8.0        |
| tty-screen                      | 0.8.1        |
| tty-cursor                      | 0.7.1        |
| tty-spinner                     | 0.9.3        |
| word_wrap                       | 1.0.0        |
| nanaimo                         | 0.3.0        |
| xcodeproj                       | 1.22.0       |
| rouge                           | 2.0.7        |
| xcpretty                        | 0.3.0        |
| xcpretty-travis-formatter       | 1.0.1        |
</details>

*generated on:* **2023-04-26**
</details>

**Comment by maksim-romanov** [comment]
Please take a look

Seems its working with `skip_waiting_for_build_processing: false` key

**Comment by axi92** [comment]
`skip_waiting_for_build_processing` is default to false according to the docs https://docs.fastlane.tools/actions/pilot/

I did not set it in my Fastfile so it should be already on false, or did I miss something?

**Comment by axi92** [comment]
I found the issue, its on apple.

TLDR: Submit your External Testing Track and let apple review and approve it, then it will work.

When you don*t have the External Testing Track active and the most important part "reviewed" it wont let you set the Groups via fastlane/apple-api.
Only when you let apple review the submission for the External Testing Track and its done and approved, the groups will then be applied on the Internal Testing tracks.

**Comment by JowyAtreides** [comment]
Hello @axi92!
I've run into the same problem described here where we want to distribute builds to an external group via the fastlane API.

What do you mean by "External Testing Track" being active on App Store Connect? In Apps/{AppName}/Testflight, the groups are divided by "Internal Testing" and "External Testing", but I don't see any option of adding any groups to review, only the builds inside the groups themselves.

With External groups that have had at least one build reviewed and submitted, distribute_external still doesn't work for me.

Maybe I'm looking in the wrong place and I would be very thankful for any help. =)

**Comment by enlivn** [comment]
@JowyAtreides I think what they mean is you must submit at least one build to the external group and (because that goes to Apple for review at least the very first time) wait for Apple to approve it.

However, I have done that and I still can't get `pilot` to send new builds to the external group with `distribute_external` set to `true`.

> Hello @axi92! I've run into the same problem described here where we want to distribute builds to an external group via the fastlane API.
> 
> What do you mean by "External Testing Track" being active on App Store Connect? In Apps/{AppName}/Testflight, the groups are divided by "Internal Testing" and "External Testing", but I don't see any option of adding any groups to review, only the builds inside the groups themselves.
> 
> With External groups that have had at least one build reviewed and submitted, distribute_external still doesn't work for me.
> 
> Maybe I'm looking in the wrong place and I would be very thankful for any help. =)

**Comment by enlivn** [comment]
@JowyAtreides okay, I finally got it to work by renaming my groups to be just single words. A space in the names causes it to not work. I don't have the time to dig into the cause but you could add prints to your _local_ `pilot` gem's `build_manager.rb` (if you use a ruby env manager, it would be at `~/.gem/ruby/3.2.2/gems/fastlane-2.225.0/pilot/lib/pilot/build_manager.rb`) here: https://github.com/fastlane/fastlane/blob/ec78d7d81a6a66e9e89fd29f6e0616d5ba09840a/pilot/lib/pilot/build_manager.rb#L460-L462, and then run the lane.

**Comment by enlivn** [comment]
> @JowyAtreides okay, I finally got it to work by renaming my groups to be just single words. A space in the names causes it to not work. I don't have the time to dig into the cause but you could add prints to your _local_ `pilot` gem's `build_manager.rb` (if you use a ruby env manager, it would be at `~/.gem/ruby/<ruby_version>/gems/fastlane-2.225.0/pilot/lib/pilot/build_manager.rb`) here:
> 
> https://github.com/fastlane/fastlane/blob/ec78d7d81a6a66e9e89fd29f6e0616d5ba09840a/pilot/lib/pilot/build_manager.rb#L460-L462
> 
> , and then run the lane.

Even more bizarre:
1. I added the prints to the location mentioned above and nothing looked wrong
2. because of 1, I reverted the name change and went back to `Internal Testers` and `External Testers`
3. I tried `pilot` again and this time it worked 😕
